specutils/analysis/resample.py

```diff
@@ -0,0 +1,219 @@
+        The resampling matrix to be applied to data arrays.
+    """
+    # Ensure the input and output dispersions are the same unit
+    with u.set_enabled_equivalencies(u.equivalencies.spectral()):
+        orig_lamb = u.Quantity(orig_lamb, unit=unit)
+        fin_lamb = u.Quantity(fin_lamb, unit=unit)
+
+        orig_lamb = orig_lamb.to(fin_lamb.unit)
+
+    orig_lamb, fin_lamb = orig_lamb.value, fin_lamb.value
+
+    orig_space = orig_lamb[1:] - orig_lamb[:-1]
+    fin_space = fin_lamb[1:] - fin_lamb[:-1]
+
+    if np.allclose(orig_space, orig_space[0]) and np.allclose(fin_space, fin_space[0]):
+        logging.info("Re-sampling: original and final grids are uniform.")
+        mat = _uniform_matrix(orig_lamb, fin_lamb)
+    else:
+        logging.info("Re-sampling: original and final grids are non-uniform.")
+        mat = _nonuniform_matrix(orig_lamb, fin_lamb, **kwargs)
+
+    return mat
+
+
+def _uniform_matrix(orig_lamb, fin_lamb):
+    """
+    Create a re-sampling matrix to be used in re-sampling spectra in a way
+    that conserves flux. This is adapted from code created by the SEAGal
+    Group.
+
+    .. note:: This method assumes uniform grids.
+
+    Parameters
+    ----------
+    orig_lamb : ndarray
+        The original dispersion array.
+    fin_lamb : ndarray
+        The desired dispersion array.
+
+    Returns
+    -------
+    resample_mat : ndarray
+        An [[N_{fin_lamb}, M_{orig_lamb}]] matrix.
+    """
+
+    # Get step size
+    delta_orig = orig_lamb[1] - orig_lamb[0]
+    delta_fin = fin_lamb[1] - fin_lamb[0]
+
+    n_orig_lamb = len(orig_lamb)
+    n_fin_lamb = len(fin_lamb)
+
+    # Lower bin and upper bin edges
+    orig_low = orig_lamb - delta_orig * 0.5
+    orig_upp = orig_lamb + delta_orig * 0.5
+    fin_low = fin_lamb - delta_fin * 0.5
+    fin_upp = fin_lamb + delta_fin * 0.5
+
+    # Create re-sampling matrix
+    resamp_mat = np.zeros(shape=(n_fin_lamb, n_orig_lamb))
+
+    for i in range(n_fin_lamb):
+        # Calculate the contribution of each original bin to the
+        # resampled bin
+        l_inf = np.where(orig_low > fin_low[i], orig_low, fin_low[i])
+        l_sup = np.where(orig_upp < fin_upp[i], orig_upp, fin_upp[i])
+
+        # Interval overlap of each original bin for current resampled
+        # bin; negatives clipped
+        dl = (l_sup - l_inf).clip(0)
+
+        # This will only happen at the edges of lorig.
+        # Discard resampled bin if it's not fully covered (> 99%) by the
+        #  original bin -- only happens at the edges of the original bins
+        if 0 < dl.sum() < 0.99 * delta_fin:
+            dl = 0 * orig_lamb
+
+        resamp_mat[i, :] = dl
+
+    resamp_mat = resamp_mat / delta_fin
+
+    return resamp_mat
+
+
+def _nonuniform_matrix(orig_lamb, fin_lamb, extrapolate=False):
+    """
+    Compute re-sampling matrix R_o2r, useful to convert a spectrum sampled at
+    wavelengths lorig to a new grid lresamp. Here, there is no necessity to
```

Comment:
there are variables referred to here that are not in the code.  Presumably `lorig`->`orig_lamb`?

Reply:
Good catch.